scarb/src/compiler/plugin/proc_macro/compilation.rs

```diff
@@ -61,6 +67,31 @@ impl SharedLibraryProvider for Package {
             .path_unchecked()
             .join(lib_name))
     }
+
+    fn prebuilt_lib_path(&self) -> Option<Utf8PathBuf> {
+        let target_triple = target!();
+
+        let prebuilt_name = format!(
+            "{name}_v{version}_{target}{suffix}",
+            name = self.id.name,
+            version = self.id.version,
+            target = target_triple,
+            suffix = DLL_SUFFIX
```

Comment:
I've considered this as well. Unfortunately, it will lead to naming conflicts, as for example library with a name `libsome_macro.so` will be produced both for `aarch64-unknown-linux-gnu` and `x86_64-unknown-linux-gnu`.

Therefore I used format as described in the issue:

> Within this directory, binaries for all platforms will follow the following naming pattern: `${PACKAGE_NAME}_v${PACKAGE_VERSION}_${TARGET}.${so|dylib|dll}`

On the second look though
```
some_macro_v0.1.0_aarch64-apple-darwin.dylib
some_macro_v0.1.0_aarch64-unknown-linux-gnu.so
some_macro_v0.1.0_x86_64-apple-darwin.dylib
some_macro_v0.1.0_x86_64-unknown-linux-gnu.so
```
we may consider dropping `${PACKAGE_NAME}_v${PACKAGE_VERSION}_` to slightly simplify these.

Reply:
I just meant replace `DLL_SUFFIX` with call to `libloading::library_filename` which adds the extension for you

Reply:
It also adds `DLL_PREFIX`, which can be either `""` or `lib`, which seems more confusing than helpful, especially since users need to name the files manually.

What do you think about dropping the `${PACKAGE_NAME}_v${PACKAGE_VERSION}_` prefix though? It would simplify things and prevent naming issues caused by mismatched Cargo and Scarb names or versions. The downside is that users might accidentally reuse a prebuilt binary after a version update, but this seems like a minor issue to me.

Reply:
> It also adds DLL_PREFIX, which can be either "" or lib, which seems more confusing than helpful, especially since users need to name the files manually.

This is what Cargo compiles this packages into. Forcing macro authors to remove the prefix from compiled file is imho more confusing, that sticking to the conventional naming. 

> naming issues caused by mismatched Cargo and Scarb names or versions

What do you mean?

Reply:
> What do you mean?

I’m referring to the reason behind [this warning](https://github.com/software-mansion/scarb/blob/31b99eccd0c6f8142d7f9069bf6152a4926d8a18/scarb/src/compiler/plugin/proc_macro/compilation.rs#L109).  

When naming build artifacts, Cargo uses the package name specified in the Cargo manifest, which:  
1. May differ from the name specified in the Scarb manifest: https://github.com/software-mansion/scarb/pull/1605#discussion_r1780213218
2. Must be resolved manually via `cargo_metadata`: https://github.com/software-mansion/scarb/pull/1605#discussion_r1786054025

The problem with 1. is that users will have to name packages manually, leading to confusion about whether the package name from the Cargo or Scarb manifest should be used - mistakes wills surely be made.  
The problem with 2. is that using proc macros *without* Cargo is one of the motivations for introducing loading prebuilts. Relying on `cargo_metadata` would make that impossible.  

> This is what Cargo compiles this packages into. Forcing macro authors to remove the prefix from compiled file is imho more confusing, that sticking to the conventional naming.

While it’s true that Cargo compiles libraries with a prefix, the resulting libraries must be renamed anyway to avoid conflicts:  
```
libsome_macro.dylib  
libsome_macro.so  
libsome_macro.dylib  
libsome_macro.so  
some_macro.dll  
```  

Since renaming is inevitable, I argue that the simplest and most consistent naming convention - requiring minimal amount of documentation reading from the user - is:
```
aarch64-apple-darwin.dylib  
aarch64-unknown-linux-gnu.so  
x86_64-apple-darwin.dylib  
x86_64-unknown-linux-gnu.so  
x86_64-pc-windows-msvc.so  
```  

This approach seems preferable to me alternatives like:  
```
libsome_macro_v0.1.0_aarch64-apple-darwin.dylib  
libsome_macro_v0.1.0_aarch64-unknown-linux-gnu.so  
libsome_macro_v0.1.0_x86_64-apple-darwin.dylib  
libsome_macro_v0.1.0_x86_64-unknown-linux-gnu.so  
some_macro_v0.1.0_x86_64-pc-windows-msvc.so  
```  

or: 
```
libaarch64-apple-darwin.dylib  
libaarch64-unknown-linux-gnu.so  
libx86_64-apple-darwin.dylib  
libx86_64-unknown-linux-gnu.so  
x86_64-pc-windows-msvc.dll  
```  

What do you think?

Reply:
> The problem with 1. is that users will have to name packages manually, leading to confusion about whether the package name from the Cargo or Scarb manifest should be used - mistakes wills surely be made.

True! Sounds like we should turn this warning into a hard error soon, to warn macro authors early on about the mismatch 😄 

> While it’s true that Cargo compiles libraries with a prefix, the resulting libraries must be renamed anyway to avoid conflicts:

Makes sense. So let's drop the prefix indeed. I would still keep the name + version, just to make sure.

Reply:
> So let's drop the prefix indeed.

👍🏻 

> I would still keep the name + version, just to make sure.

The current version does so, but I feel strongly that we should drop the name and version as well. They're redundant since this info is already available in the manifest, tarball name, and elsewhere, which potentially just makes things just a bit harder fot the macro developers. The only potential benefit I see is reducing the likelihood of outdated binaries being included. Unless you have any other reasons to keep it, I don't think this downside is really that important and justifies the trade-off.